Source/Tests/LinuxSyscalls/Syscalls.cpp

```diff
@@ -226,89 +211,156 @@ static bool IsSupportedByInterpreter(std::string const &Filename) {
   return false;
 }
 
-uint64_t ExecveHandler(const char *pathname, char* const* argv, char* const* envp, ExecveAtArgs *Args) {
+static bool IsShebangFilename(std::string const &Filename) {
+  // Open the Filename to determine if it is a shebang file.
+  std::fstream File;
+  size_t FileSize{0};
+  File.open(Filename, std::fstream::in | std::fstream::binary);
+
+  if (!File.is_open())
+    return false;
+
+  File.seekg(0, File.end);
+  FileSize = File.tellg();
+  File.seekg(0, File.beg);
+
+  // Is the file large enough for shebang
+  if (FileSize <= 2)
+    return false;
+
+  // The maximum length of the shebang line is `#!` + 255 chars
+  std::vector<char> Header(std::min(257ul, FileSize));
+  File.read(&Header.at(0), Header.size());
+
+  return IsShebangFile(&Header.at(0), Header.size());
+}
+
+static bool IsShebangFD(int FD) {
+  // We don't know the state of the FD coming in since this might be a guest tracked FD.
+  // Need to be extra careful here not to adjust file offsets and status flags.
+  //
+  // Can't use dup since that makes the FD have the same file description backing both FDs.
+
+  // We need to first determine the file size through fstat.
+  struct stat buf{};
+  if (fstat(FD, &buf) == -1) {
+    // Couldn't get size.
+    return false;
+  }
+
+  // Is the file large enough for shebang
+  if (buf.st_size <= 2)
+    return false;
+
+  // The maximum length of the shebang line is `#!` + 255 chars
+  std::vector<char> Header(std::min(257l, buf.st_size));
+
+  // Read the header so we can tell if it is a supported ELF file.
+  // Can't adjust file offset, so use pread.
+  if (pread(FD, &Header.at(0), Header.size(), 0) != Header.size()) {
+    // Couldn't read
+    LogMan::Msg::EFmt("Couldn't read potential shebang FD");
+    return false;
+  }
+
+  return IsShebangFile(&Header.at(0), Header.size());
+}
+
+uint64_t ExecveHandler(const char *pathname, char* const* argv, char* const* envp, ExecveAtArgs &Args) {
   std::string Filename{};
 
   std::error_code ec;
   std::string RootFS = FEX::HLE::_SyscallHandler->RootFSPath();
-
-  // Check the rootfs if it is available first
-  if (pathname[0] == '/') {
-    auto Path = FEX::HLE::_SyscallHandler->FM.GetEmulatedPath(pathname, true);
-    if (!Path.empty() && std::filesystem::exists(Path, ec)) {
-      Filename = Path;
+  ELFLoader::ELFContainer::ELFType Type{};
+  bool IsFDExec{};
```

Comment:
```suggestion
  const bool IsFDExec = (Args.flags & AT_EMPTY_PATH) && strlen(pathname) == 0;
```
(Keeps state a bit more local)

Reply:
Resolved.